### PR TITLE
Issue 8538

### PR DESCRIPTION
**Pull request by erickgonzalez**
Before we only allowed CMS Admin to create top-level categories, now if a user has the required permissions (Add Children and Edit at All Sites) is able to add top-level categories.

Also a small refactor to the CategoryAPI was made, unused methods were deleted, even permissions level and logs.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java [hunk 1]
```diff
@@ -142,41 +102,55 @@ public Category find(final String id, final User user,
...
+	 * @param parent Parent can be null if saving an top level category
+	 * @param category Category to be saved
+	 * @param user user that is performing the save
+	 * @throws DotDataException
+	 * @throws DotSecurityException
+	 */
 	@WrapInTransaction
 	public void save(final Category parent,
-					 Category object,
+					 Category category,
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java:120]: Codacy found an issue: [Avoid reassigning parameters such as 'category'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6178451)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/categories/business/CategoryAPITest.java [hunk 2]
```diff
@@ -1084,4 +1077,190 @@ private ContentType createContentTypeWithCatAndTextField(final Category parentCa
...
+    @Test
+    public void test_save_createSubCategory_asAdmin_success()
+            throws DotSecurityException, DotDataException {
+        final String parentCategoryName = "newParentCategory" + System.currentTimeMillis();
+        final String childCategoryName = "newChildCategory" + System.currentTimeMillis();
+        //Create Child Category
+        final Category newChildCategory = new CategoryDataGen().setCategoryName(childCategoryName)
+                .setCategoryVelocityVarName(childCategoryName).setKey(childCategoryName).next();
+        //Create Parent Category
+        final Category newParentCategory = new CategoryDataGen().setCategoryName(parentCategoryName)
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/categories/business/CategoryAPITest.java:1114]: Codacy found an issue: [Avoid unused local variables such as 'newParentCategory'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6178451)

File: dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java [hunk 3]
```diff
@@ -142,41 +102,55 @@ public Category find(final String id, final User user,
...
-			 }
-		} else {
-			//Object is new and a parent was provided so we check in the parent permissions
-			if(!permissionAPI.doesUserHavePermission(parent, PermissionAPI.PERMISSION_EDIT_PERMISSIONS, user, respectFrontendRoles))
-				throw new DotSecurityException("User doesn't have permission to save this category = " +
-						object.getInode() + " having as parent the category = " + parent.getInode());
+	    category = checkUniqueKey(category, user);
 
-			isANewCategory = true;
+	    boolean isANewCategory = UtilMethods.isNotSet(category.getInode());
```
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java:127]: Codacy found an issue: [Avoid declaring a variable if it is unreferenced before a possible exit point.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6178451)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/categories/business/CategoryAPITest.java [hunk 4]
```diff
@@ -1084,4 +1077,190 @@ private ContentType createContentTypeWithCatAndTextField(final Category parentCa
...
+     * Method to test: {@link CategoryAPI#save(Category, Category, User, boolean)}
+     * Given Scenario: As a admin user create a top level category
+     * ExpectedResult: Category created successfully
+     */
+    @Test
+    public void test_save_createTopLevelCategory_asAdmin_success()
+            throws DotSecurityException, DotDataException {
+        //Create new Top Level Category
+        final String categoryName = "newCategory" + System.currentTimeMillis();
+        final Category newCategory = new CategoryDataGen().setCategoryName(categoryName)
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/categories/business/CategoryAPITest.java:1091]: Codacy found an issue: [Avoid unused local variables such as 'newCategory'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6178451)

File: dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java [hunk 5]
```diff
@@ -142,41 +102,55 @@ public Category find(final String id, final User user,
...
+
+	    //If parent is null is a top level category, we need to check permissions over the SYSTEM_HOST
+		if(!UtilMethods.isSet(parent)){
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java:130]: maybe it should go to a local method, something like:

```
private boolean isTopLevelCategory(parent){
    return !UtilMethods.isSet(parent);
}
```

File: dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java [hunk 6]
```diff
@@ -142,41 +102,55 @@ public Category find(final String id, final User user,
...
+				if(!permissionAPI.doesUserHavePermission(APILocator.systemHost(), PermissionAPI.PERMISSION_CAN_ADD_CHILDREN, user, respectFrontendRoles) ||
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java:131]: maybe it should go to a local method too, something like:

```
private boolean checkTopLevelCategoryPermission(parent, user, respectFrontendRoles ){
    return !UtilMethods.isSet(parent);
}
```

File: dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java [hunk 7]
```diff
@@ -142,41 +102,55 @@ public Category find(final String id, final User user,
...
+						!permissionAPI.doesUserHavePermissions(PermissionableType.CATEGORY, PermissionAPI.PERMISSION_EDIT, user)){
+					Logger.error(this,"User doesn't have permission to save the category at top level");
+				throw new DotSecurityException(
+						"User doesn't have permission to save the category at top level");
+			}
+		} else {
+			if (!permissionAPI.doesUserHavePermission(parent, PermissionAPI.PERMISSION_EDIT, user, respectFrontendRoles)) {
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java:138]: maybe it should go to a local method too, something like:

```
private boolean checkCategoryPermission(parent, user, respectFrontendRoles ){
    return !UtilMethods.isSet(parent);
}
```

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java (hunk 1 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java:120]: Codacy found an issue: [Avoid reassigning parameters such as 'category'](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6178451)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/categories/business/CategoryAPITest.java [hunk 8]
```diff
@@ -1087,6 +1080,192 @@ private ContentType createContentTypeWithCatAndTextField(final Category parentCa
...
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/categories/business/CategoryAPITest.java:1092]: Codacy found an issue: [Avoid unused local variables such as 'newCategory'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6178451)

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/categories/business/CategoryAPITest.java [hunk 9]
```diff
@@ -1087,6 +1080,192 @@ private ContentType createContentTypeWithCatAndTextField(final Category parentCa
...
+    @Test
+    public void test_save_createSubCategory_asAdmin_success()
+            throws DotSecurityException, DotDataException {
+        final String parentCategoryName = "newParentCategory" + System.currentTimeMillis();
+        final String childCategoryName = "newChildCategory" + System.currentTimeMillis();
+        //Create Child Category
+        final Category newChildCategory = new CategoryDataGen().setCategoryName(childCategoryName)
+                .setCategoryVelocityVarName(childCategoryName).setKey(childCategoryName).next();
+        //Create Parent Category
+        final Category newParentCategory = new CategoryDataGen().setCategoryName(parentCategoryName)
```
Comment (dev-dotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/categories/business/CategoryAPITest.java:1115]: Codacy found an issue: [Avoid unused local variables such as 'newParentCategory'.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6178451)

File: dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java (hunk 3 above)
Comment (dev-dotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/categories/business/CategoryAPIImpl.java:127]: Codacy found an issue: [Avoid declaring a variable if it is unreferenced before a possible exit point.](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=6178451)

**Review by jdotcms — APPROVED** [review]

**Pull request merged**